[Detection Engine] Adds alternate path/mappings for source data_stream fields in alerts index

rylnd opened the pull request:
## Summary

This PR addresses https://github.com/elastic/kibana/issues/171104. A summary of changes is as follows:

1. Adds new `keyword` mappings to the alerting index(es):
    * `kibana.alert.original_data_stream.dataset`
    * `kibana.alert.original_data_stream.namespace`
    * `kibana.alert.original_data_stream.type`
2. Adds new Detection Engine logic to copy these source data to the new path while building the alert
    * Formalizes/abstracts the above copying mechanism, as we were previously doing this for `event.*` fields as well.


## Steps to review
1. Sanity-check my changes to the versioned alerting types (I just cargo culted that)
1. Load some data with the relevant fields (here we use the archive accompanying these integration tests):
    
       TEST_ES_URL=http://elastic:changeme@localhost:9200 TEST_KIBANA_URL=http://elastic:changeme@localhost:5601 node scripts/es_archiver.js load x-pack/test/functional/es_archives/security_solution/ecs_fields_duplicated_for_alerts

1. Create a rule against that data:
    * index: `ecs_fields_duplicated_for_alerts`
    * 3 records ~ `2025-05-07T14:34:11.000Z`

1. Observe that the generated alerts contain the new fields referenced above:
    
    <kbd><img width="700" alt="Screenshot 2025-05-07 at 5 21 47 PM" src="https://github.com/user-attachments/assets/fc765480-c57a-4149-937c-8d8acaa082b2" /></kbd>

1. (Optional) Observe that the generated alerts contain the existing `kibana.alert.original_event.*` fields as before:
    <kbd><img width="700" alt="Screenshot 2025-05-07 at 5 21 56 PM" src="https://github.com/user-attachments/assets/da3b8396-40b8-47d6-b9dc-81861f54a7ee" /></kbd>

## Release note
Improves alert searching and filtering by including `data_stream` ECS fields under `kibana.alert.original_data_stream.*`

### Checklist

- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [x] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)



<!--ONMERGE {"backportTargets":["8.19"]} ONMERGE-->

rylnd committed:
Formalize logic around copying alert fields to new locations

We're already doing this for `event`, and we're about to do it for
specific data_stream fields.

rylnd committed:
Adds data_stream.* to our list of paths to copy in alerts

This adds some new constants (similar to event.* fields) and the new
key/value to the map so that these fields will be copied during alert
creation. Unit tests to confirm/document.

rylnd committed:
Add integration tests for ECS field duplication in alerts

This adds a source archive containing both data_stream.* and event.*
fields, and an integration suite that exercises the new duplication
logic both through manual rule execution and preview rule execution.

rylnd committed:
Define new field mappings for data_stream alerting fields

rylnd committed:
Update our versioned alerts schema with 8.19 changes

Currently, this is limited to the new data_stream.* fields as per
https://github.com/elastic/kibana/issues/171104.

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/299571)
* Commit: 9c0e34ab8018d4a162779c909dd304c10d0a24bf

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/299571#0196acdf-ba67-4a85-bf47-b0df88e23343)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299571","number":299571,"commit":"9c0e34ab8018d4a162779c909dd304c10d0a24bf"}],"number":299571}
buildkite-pr-comment-->

rylnd committed:
Fix version reference in comment

rylnd committed:
Fix Alert model definition for 8.19

I made a mess of this on the first attempt, and then discovered what
appear to be mistakes/inconsistencies in the 8.18 definition. While I
sort that out, this should fix the majority of type errors that the bad
definition caused.

rylnd committed:
Temporarily satisfy the type issue and add a TODO to discuss

This doesn't feel like the right solution here, but I need more context
on the intent behind this mechanism.

rylnd commented:
/ci

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/299598)
* Commit: e3dea8df9ff3495d594f09d780686405b91adc85

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/299598#0196ae18-ca2e-4673-8de7-af7e72fc47cc)

### History
* :broken_heart: [Build #299571](https://buildkite.com/elastic/kibana-pull-request/builds/299571) failed 9c0e34ab8018d4a162779c909dd304c10d0a24bf

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299598","number":299598,"commit":"e3dea8df9ff3495d594f09d780686405b91adc85"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299571","number":299571,"commit":"9c0e34ab8018d4a162779c909dd304c10d0a24bf"}],"number":299598}
buildkite-pr-comment-->

rylnd commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/rule_types/factories/utils/build_alert.ts:
```diff
@@ -279,6 +282,10 @@ export const buildAlertFields = ({
     [ALERT_RULE_VERSION]: params.version,
     [ALERT_URL]: alertUrl,
     [ALERT_UUID]: alertUuid,
+    // TODO this satisfies the types, but we always generate these fields in `transformAndMutateDoc`, as opposed to here
```
@marshallmain can you weigh in on this change? This indicates to me that I defined something wrong in the 8.19 alert model, but I don't know enough about how that works to confidently make the change. This and the [other](https://buildkite.com/elastic/kibana-pull-request/builds/299598#0196ae18-ca2e-4673-8de7-af7e72fc47cc) type failure are currently due to the alert model requiring these new fields in certain areas.

It seems like the obvious answer is to make these fields _optional_ in the alert model, but I didn't see any examples of that so wanted to discuss with you first.

marshallmain replied:
I think the "right" way to do this would be to separate the return type of `buildAlert` from the return type of `transformHitToAlert` - right now they're both `BaseFieldsLatest`, but `transformHitToAlert` adds extra fields. `buildAlert` fits the "base" concept better IMO, so we'd need another term for the return type of `transformHitToAlert`. `FullAlert`? `TransformedAlert`? And we'd add the original data stream (and original event fields, which we don't have in the model right now) to this new extended type.

Since we don't have original event fields properly in the schema yet, and these alert schema types aren't providing as much value as I hoped vs the maintenance cost when I set up this structure...maybe we can just omit these new ones for now? Unless you have some time to collab on improving the way we set up all these alert schema types by version.

rylnd replied:
I was using https://github.com/elastic/kibana/pull/206582 as reference for my PR, and it looks like that one made the choice to add their new fields as `undefined` in those two places (as I've now done in one of them) to sidestep the issue. So if we did that, we'd only be slightly worse off than we are now 😃 .

RE the value/efficacy of these schemae: it was tricky to implement a new one, even with documentation; I think the most onerous part is having to re-export stuff that you didn't change (i.e. the change here to the "base" fields should just flow to the other types).

The types are also not fully accurate right now, with e.g. no reference to `original_event` (as you said). I also believe some of those models were not updated correctly, e.g. [this definition](https://github.com/elastic/kibana/pull/206582/files#diff-74d64034c85059189032c81328368c81ff7bbc004127ce65a038c4fc2e81690bR54) should be :

```diff
- export type EqlBuildingBlockAlert8180 = AlertWithCommonFields800<EqlBuildingBlockFields8160>;
+ export type EqlBuildingBlockAlert8180 = AlertWithCommonFields800<EqlBuildingBlockFields8180>;
```

so: confidence is low.

I'm gonna noodle on a better solution overall, and in the meantime just add those fields to the other place to get past the type errors.

rylnd replied:
@marshallmain I do have one more question in the meantime: what is the _meaning_ behind e.g. `BaseFieldsLatest`? The accompanying comment states that it represents "alerts that shipped in [kibana version]," which is quite vague.

* If it's meant to represent the superset of all possible alert fields, then it shouldn't be used as the return type for our "base field builder" utilities
* If it's meant to represent a guaranteed subset of alert fields, then it shouldn't include these new fields (nor should it include those introduced in #206582, at the very least).

marshallmain commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/rule_types/factories/utils/traverse_and_mutate_doc.ts:
```diff
@@ -305,3 +306,47 @@ const traverseArray = ({
+ */
+const pathNeedsCopying = (fullPath: string): boolean => {
+  const topLevelPath = getTopLevelPath(fullPath);
+  if (!topLevelPath) {
```
can `topLevelPath` ever be undefined?

rylnd replied:
If we somehow received a `fullPath` of `""` (empty string) or `".*"` (something that starts with a period, we'd receive an empty string for `topLevelPath`. Typescript is telling me `fullPath` is a string, though, so `topLevelPath` will never be undefined.

I think this all flows as expected, but I can write a couple quick unit tests for those cases.

rylnd replied:
@marshallmain this code is now outdated, but FYI I did write some tests for this as I was curious what the code would do. I'm not committing it because it doesn't seem useful, but here's what they look like: 


```ts
    describe('edge cases', () => {
      it('preserves an empty nested top-level field', () => {
        const { result, removed } = traverseAndMutateDoc({
          '': { foo: 'test-foo', type: 'test-type' },
        });

        expect(result).toEqual({
          '': { foo: 'test-foo', type: 'test-type' },
        });

        expect(removed).toEqual([]);
      });

      it('drops an empty dotted top-level field', () => {
        const { result, removed } = traverseAndMutateDoc({
          '.foo': 'test-foo',
          '.type': 'test-type',
        });

        expect(result).toEqual({
          '.foo': 'test-foo',
          '.type': 'test-type',
        });
        expect(removed).toEqual([]);
      });
    });
  });
```

rylnd commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/rule_types/factories/utils/traverse_and_mutate_doc.ts:
```diff
@@ -305,3 +306,47 @@ const traverseArray = ({
+    alertingNamespaceCopyMap[topLevelPath as keyof typeof alertingNamespaceCopyMap];
+
+  if (!copyPathRoot) {
+    throw new Error(`No copy destination path found for path '${fullPath}'`);
```
Yeah, this is basically unreachable code; the `pathNeedsCopying` guard ensures that we're only calling this when we should, and copyPathRoot will be defined. I'll take another pass on this to remove the throwing and instead just not call as you suggest.

rylnd replied:
Fixed in https://github.com/elastic/kibana/pull/220447/commits/4f289eab1791c6f75c9344656751627f1104f9e2.

rylnd committed:
Satisfy types temporarily

These will need to be fixed before merging, but for now this gets us
further in CI.

rylnd commented:
/ci

rylnd committed:
Merge branch 'main' into rylnd/alert_datastream_fields

rylnd committed:
Remove unnecessary type guards (and a throw) with better Typescript

* Adds a predicate function to narrow the type to a key of the map that
  we're indexing into
* Generally fixes the flow to better represent our outer guards; we will
  neither have an empty topLevelPath (by merit of `topLevel` being true)
  nor will we have an empty result from `getCopyDestinationPath` (by
  merit of the previous bullet)

rylnd committed:
Merge branch 'main' into rylnd/alert_datastream_fields

rylnd committed:
Revert "Satisfy types temporarily"

This reverts commit 8defebdeb8173392db3bf463edff19a7014c1137.

This is not the approach we want to take. Reverting instead of rebasing
since this branch has already had some review/conversation done.

rylnd committed:
Make new data_stream fields optional in our alert model

This is the first example of optional fields in the model, but this
allows us to both keep the existing types and represent these new fields
as accurately as possible. I will be filing an issue to address the
general issues with the model as a followup.

rylnd committed:
Merge branch 'main' into rylnd/alert_datastream_fields

rylnd committed:
Remove defaulting of fields to satisfy (now outdated) return types

Now that these fields are optional, this change is not necessary.

rylnd commented:
/ci

rylnd committed:
Update AAD snapshots with new (optional) alert fields

rylnd committed:
Update generated alerting schemas used in canary test

This was done by following the instructions in
/x-pack/test/alerting_api_integration/spaces_only/tests/alerting/group4/generate_alert_schemas.ts.

rylnd committed:
Add new fields to relevant integration tests

These tests create alerts from data_stream documents, and thus will have
these fields copied over (as expected by this branch).

rylnd commented:
/ci

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/303545)
* Commit: 7a57fc648991ba1c79c5ddb589957e025e3fab44

### Failed CI Steps
* [FTR Configs #40](https://buildkite.com/elastic/kibana-pull-request/builds/303545#01971e48-729a-4855-a558-15d0342b8860)
* [FTR Configs #40](https://buildkite.com/elastic/kibana-pull-request/builds/303545#01971e71-e79b-49bc-9cf1-028bd9434aef)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303545#01971e48-729a-4855-a558-15d0342b8860) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303545/jobs/01971e48-729a-4855-a558-15d0342b8860/artifacts/01971e71-a389-4865-af32-003acb0c8e03) FTR Configs #<span></span>40 / Embeddable Alerts Table Embeddable alerts panel "before all" hook for "should only show alerts from the observability area (o11y+stack) when selecting it"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303545#01971e71-e79b-49bc-9cf1-028bd9434aef) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303545/jobs/01971e71-e79b-49bc-9cf1-028bd9434aef/artifacts/01971e7a-ac89-4756-83ae-d421a72d84bd) FTR Configs #<span></span>40 / Embeddable Alerts Table Embeddable alerts panel "before all" hook for "should only show alerts from the observability area (o11y+stack) when selecting it"

### History
* :broken_heart: [Build #303522](https://buildkite.com/elastic/kibana-pull-request/builds/303522) failed 9a5abf76292fcf68ad911279b7a7ff385433e0f5
* :broken_heart: [Build #303502](https://buildkite.com/elastic/kibana-pull-request/builds/303502) failed 80a9577bcf4ef7258b87e891170d45b882294a98
* :broken_heart: [Build #299899](https://buildkite.com/elastic/kibana-pull-request/builds/299899) failed 8defebdeb8173392db3bf463edff19a7014c1137
* :broken_heart: [Build #299598](https://buildkite.com/elastic/kibana-pull-request/builds/299598) failed e3dea8df9ff3495d594f09d780686405b91adc85
* :broken_heart: [Build #299571](https://buildkite.com/elastic/kibana-pull-request/builds/299571) failed 9c0e34ab8018d4a162779c909dd304c10d0a24bf

cc @rylnd

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303545","number":303545,"commit":"7a57fc648991ba1c79c5ddb589957e025e3fab44"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303522","number":303522,"commit":"9a5abf76292fcf68ad911279b7a7ff385433e0f5"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303502","number":303502,"commit":"80a9577bcf4ef7258b87e891170d45b882294a98"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299899","number":299899,"commit":"8defebdeb8173392db3bf463edff19a7014c1137"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299598","number":299598,"commit":"e3dea8df9ff3495d594f09d780686405b91adc85"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299571","number":299571,"commit":"9c0e34ab8018d4a162779c909dd304c10d0a24bf"}],"number":303545}
buildkite-pr-comment-->

rylnd committed:
Merge branch 'main' into rylnd/alert_datastream_fields

elasticmachine commented:
Pinging @elastic/security-detection-engine (Team:Detection Engine)

rylnd committed:
Merge branch 'main' into rylnd/alert_datastream_fields

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/304925)
* Commit: 4fd8b0cd657958deccfb449035227bb0ea32231b

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :yellow_heart: [Build #303562](https://buildkite.com/elastic/kibana-pull-request/builds/303562) was flaky 3aad39f94c54eada78d528e3320ced224ec573f9
* :broken_heart: [Build #303545](https://buildkite.com/elastic/kibana-pull-request/builds/303545) failed 7a57fc648991ba1c79c5ddb589957e025e3fab44
* :broken_heart: [Build #303522](https://buildkite.com/elastic/kibana-pull-request/builds/303522) failed 9a5abf76292fcf68ad911279b7a7ff385433e0f5
* :broken_heart: [Build #303502](https://buildkite.com/elastic/kibana-pull-request/builds/303502) failed 80a9577bcf4ef7258b87e891170d45b882294a98
* :broken_heart: [Build #299899](https://buildkite.com/elastic/kibana-pull-request/builds/299899) failed 8defebdeb8173392db3bf463edff19a7014c1137
* :broken_heart: [Build #299598](https://buildkite.com/elastic/kibana-pull-request/builds/299598) failed e3dea8df9ff3495d594f09d780686405b91adc85

cc @rylnd

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/304925","number":304925,"commit":"4fd8b0cd657958deccfb449035227bb0ea32231b"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303562","number":303562,"commit":"3aad39f94c54eada78d528e3320ced224ec573f9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303545","number":303545,"commit":"7a57fc648991ba1c79c5ddb589957e025e3fab44"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303522","number":303522,"commit":"9a5abf76292fcf68ad911279b7a7ff385433e0f5"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303502","number":303502,"commit":"80a9577bcf4ef7258b87e891170d45b882294a98"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299899","number":299899,"commit":"8defebdeb8173392db3bf463edff19a7014c1137"}],"number":304925}
buildkite-pr-comment-->

rylnd committed:
Merge branch 'main' into rylnd/alert_datastream_fields

rylnd committed:
Merge branch 'main' into rylnd/alert_datastream_fields

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/307460)
* Commit: e359a1790537fc713252c477c1c1a471980b81d7

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #306421](https://buildkite.com/elastic/kibana-pull-request/builds/306421) succeeded 57e3faefa33add19970cb04a8a07a681079fadfb
* :green_heart: [Build #304925](https://buildkite.com/elastic/kibana-pull-request/builds/304925) succeeded 4fd8b0cd657958deccfb449035227bb0ea32231b
* :yellow_heart: [Build #303562](https://buildkite.com/elastic/kibana-pull-request/builds/303562) was flaky 3aad39f94c54eada78d528e3320ced224ec573f9
* :broken_heart: [Build #303545](https://buildkite.com/elastic/kibana-pull-request/builds/303545) failed 7a57fc648991ba1c79c5ddb589957e025e3fab44
* :broken_heart: [Build #303522](https://buildkite.com/elastic/kibana-pull-request/builds/303522) failed 9a5abf76292fcf68ad911279b7a7ff385433e0f5
* :broken_heart: [Build #303502](https://buildkite.com/elastic/kibana-pull-request/builds/303502) failed 80a9577bcf4ef7258b87e891170d45b882294a98

cc @rylnd

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307460","number":307460,"commit":"e359a1790537fc713252c477c1c1a471980b81d7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306421","number":306421,"commit":"57e3faefa33add19970cb04a8a07a681079fadfb"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/304925","number":304925,"commit":"4fd8b0cd657958deccfb449035227bb0ea32231b"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303562","number":303562,"commit":"3aad39f94c54eada78d528e3320ced224ec573f9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303545","number":303545,"commit":"7a57fc648991ba1c79c5ddb589957e025e3fab44"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303522","number":303522,"commit":"9a5abf76292fcf68ad911279b7a7ff385433e0f5"}],"number":307460}
buildkite-pr-comment-->

pmuellr reviewed (approved):
ResponseOps changes LGTM

marshallmain reviewed (approved):
LGTM, thanks for extracting the logic around copying certain paths to be more reusable 🚀

marshallmain commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/general_logic/basic_license_essentials_tier/ecs_field_duplication.ts:
```diff
@@ -0,0 +1,99 @@
+        const { id } = await createRule(supertest, log, rule);
+        await waitForRuleSuccess({ supertest, log, id });
+        await waitForAlertsToBePresent(supertest, log, 3, [id]);
+        const alertsResponse = await getAlertsById(supertest, log, id);
```
might be useful to test filtering by the new `data_stream` fields as well to make sure they get mapped correctly?

rylnd replied:
Good idea, added just such a test in https://github.com/elastic/kibana/pull/220447/commits/cc72a1b38d3c2c6f95ff177a6e6ab0f0ddced31a  👍

rylnd committed:
Adds an integration test to verify the new data_stream mappings

In this test we create the rule/alert as we had in previous tests, but
instead of asserting on the contained fields, we first _retrieve_ the
alerts by searching for them by these new fields. Since the query uses a
boolean `must`, we know that the returned alerts contain all the fields
specified in the query.

Since we didn't have a nice wrapper for performing a general query to
our "search alerts" API, I created one here (`searchAlerts`).

rylnd commented on x-pack/test/functional/es_archives/security_solution/ecs_fields_duplicated_for_alerts/data.json:
```diff
@@ -0,0 +1,44 @@
+  "value": {
+    "id": "1",
+    "index": "ecs_fields_duplicated_for_alerts",
+    "source": {
```
FYI this new archive reads a little wonky since you'd expect the `data_stream` fields to be in here, right? But because they're `constant_keyword` fields they are defined on the mappings themselves, and not the documents.

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/308098)
* Commit: cc72a1b38d3c2c6f95ff177a6e6ab0f0ddced31a

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/308098#01976602-02ca-44b2-8381-0a53b053f729)

### History
* :green_heart: [Build #307460](https://buildkite.com/elastic/kibana-pull-request/builds/307460) succeeded e359a1790537fc713252c477c1c1a471980b81d7
* :green_heart: [Build #306421](https://buildkite.com/elastic/kibana-pull-request/builds/306421) succeeded 57e3faefa33add19970cb04a8a07a681079fadfb
* :green_heart: [Build #304925](https://buildkite.com/elastic/kibana-pull-request/builds/304925) succeeded 4fd8b0cd657958deccfb449035227bb0ea32231b
* :yellow_heart: [Build #303562](https://buildkite.com/elastic/kibana-pull-request/builds/303562) was flaky 3aad39f94c54eada78d528e3320ced224ec573f9
* :broken_heart: [Build #303545](https://buildkite.com/elastic/kibana-pull-request/builds/303545) failed 7a57fc648991ba1c79c5ddb589957e025e3fab44
* :broken_heart: [Build #303522](https://buildkite.com/elastic/kibana-pull-request/builds/303522) failed 9a5abf76292fcf68ad911279b7a7ff385433e0f5

cc @rylnd

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308098","number":308098,"commit":"cc72a1b38d3c2c6f95ff177a6e6ab0f0ddced31a"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307460","number":307460,"commit":"e359a1790537fc713252c477c1c1a471980b81d7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306421","number":306421,"commit":"57e3faefa33add19970cb04a8a07a681079fadfb"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/304925","number":304925,"commit":"4fd8b0cd657958deccfb449035227bb0ea32231b"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303562","number":303562,"commit":"3aad39f94c54eada78d528e3320ced224ec573f9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303545","number":303545,"commit":"7a57fc648991ba1c79c5ddb589957e025e3fab44"}],"number":308098}
buildkite-pr-comment-->

rylnd committed:
Fix type error

The search body passed to supertest needs to be `string | object | undefined`

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/308131)
* Commit: 5b8d214e6afc2c4d429d3536030c089c4298e6cb

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #308098](https://buildkite.com/elastic/kibana-pull-request/builds/308098) failed cc72a1b38d3c2c6f95ff177a6e6ab0f0ddced31a
* :green_heart: [Build #307460](https://buildkite.com/elastic/kibana-pull-request/builds/307460) succeeded e359a1790537fc713252c477c1c1a471980b81d7
* :green_heart: [Build #306421](https://buildkite.com/elastic/kibana-pull-request/builds/306421) succeeded 57e3faefa33add19970cb04a8a07a681079fadfb
* :green_heart: [Build #304925](https://buildkite.com/elastic/kibana-pull-request/builds/304925) succeeded 4fd8b0cd657958deccfb449035227bb0ea32231b
* :yellow_heart: [Build #303562](https://buildkite.com/elastic/kibana-pull-request/builds/303562) was flaky 3aad39f94c54eada78d528e3320ced224ec573f9

cc @rylnd

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308131","number":308131,"commit":"5b8d214e6afc2c4d429d3536030c089c4298e6cb"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308098","number":308098,"commit":"cc72a1b38d3c2c6f95ff177a6e6ab0f0ddced31a"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307460","number":307460,"commit":"e359a1790537fc713252c477c1c1a471980b81d7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306421","number":306421,"commit":"57e3faefa33add19970cb04a8a07a681079fadfb"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/304925","number":304925,"commit":"4fd8b0cd657958deccfb449035227bb0ea32231b"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303562","number":303562,"commit":"3aad39f94c54eada78d528e3320ced224ec573f9"}],"number":308131}
buildkite-pr-comment-->

kibanamachine commented:
Starting backport for target branches: 8.19

https://github.com/elastic/kibana/actions/runs/15623282119

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.19|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/223849">](https://github.com/elastic/kibana/pull/223849)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->

rylnd commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|9.0|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/223854">](https://github.com/elastic/kibana/pull/223854)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 10.0.0 -->

irfaan0999-alt commented:
Hello  team,

We created a custom fields on an agent policy but we noticed that the field does not get stored on the index .internal.alerts-security.alerts-default whenever an alert is triggered. Will this be resolved along with the namespace issue?
